Nonprofit-Exchange-Hub/web-app
# Esteban/integrate users with categories

## Dev Summary

This PR proposes to store the users' interests in an `interests` field of type `JSONB` instead of adding a new many-to-many table. The trade-off is that if a specific `categories` record changes, the list of interests won't change automatically. The assumption is that the user will be able to update their interests in their profiles as they wish. In that case, a new interest list can be stored in the `interests`  field.

The PR also includes
- a working service method to query the users based on a list of interests
- a new method in the categories service to validate a list of categories
- 2 new custom dto validators to validate the correct shape of the `JSONB` interests object




## Test Plan


repro steps:

1. run the app.
2. register a new user
```bash
curl --location 'http://localhost:3001/api/auth/register' \
--header 'Content-Type: application/json' \
--data-raw '{
    "firstName": "Billie",
    "last_name": "Eilish",
    "email": "user5@example.com",
    "password": "Secret1234%",
    "email_notification_opt_out": true,
    "city": "seattle",
    "state": "wa",
    "interests": {
        "names": [
            "furniture"
        ]
    }
}'
```
3. log in and expect to see a response that includes the the interests field:
```json
.....
        "interests": {
            "names": [
                "furniture"
            ]
        }
.....
```
4. validate categories endpoint should return true when the categories are valid, otherwise, false
```bash
curl --location 'http://localhost:3001/api/categories/validate' \
--header 'Content-Type: application/json' \
--header 'Cookie: NEH_is_cool={your_cookie}' \
--data '[
    "paper products",
    "furniture"
]'
```






## Resources

- [postgres JSONB with typeorm](https://wanago.io/2020/12/28/nestjs-json-postgresql-typeorm/)
